packages/std/src/ibc_exports.rs

```diff
@@ -0,0 +1,251 @@
+#![cfg(all(feature = "stargate", target_arch = "wasm32"))]
+use std::fmt;
+use std::vec::Vec;
+
+use schemars::JsonSchema;
+use serde::Serialize;
+
+use crate::deps::DepsMut;
+use crate::exports::make_dependencies;
+use crate::ibc::{IbcBasicResponse, IbcChannel};
+use crate::memory::{consume_region, release_buffer, Region};
+use crate::results::ContractResult;
+use crate::serde::{from_slice, to_vec};
+use crate::types::Env;
+use crate::{IbcAcknowledgement, IbcPacket, IbcReceiveResponse};
+
+// TODO: replace with https://doc.rust-lang.org/std/ops/trait.Try.html once stabilized
+macro_rules! r#try_into_contract_result {
+    ($expr:expr) => {
+        match $expr {
+            Ok(val) => val,
+            Err(err) => {
+                return ContractResult::Err(err.to_string());
+            }
+        }
+    };
+    ($expr:expr,) => {
+        $crate::try_into_contract_result!($expr)
+    };
+}
+
+/// do_ibc_channel_open is designed for use with #[entry_point] to make a "C" extern
+///
+/// contract_fn does the protocol version negotiation during channel handshake phase
+pub fn do_ibc_channel_open<E>(
+    contract_fn: &dyn Fn(DepsMut, Env, IbcChannel) -> Result<(), E>,
+    env_ptr: u32,
+    msg_ptr: u32,
+) -> u32
+where
+    E: ToString,
+{
+    let res = _do_ibc_channel_open(contract_fn, env_ptr as *mut Region, msg_ptr as *mut Region);
+    let v = to_vec(&res).unwrap();
+    release_buffer(v) as u32
+}
+
+fn _do_ibc_channel_open<E>(
+    contract_fn: &dyn Fn(DepsMut, Env, IbcChannel) -> Result<(), E>,
+    env_ptr: *mut Region,
+    msg_ptr: *mut Region,
+) -> ContractResult<bool>
```

Comment:
If this does not have a return value, why not `ContractResult<()>`?

Reply:
When we marshall the json and it goes into go, it is a bit annoying to deal with enums.

We usually check for the `"ok"` field, then the `"error"` field. I added `true` here, just so the `"ok"` field would have something. If someone returned `Err("")`, there would be no way for go to distinguish between ok and error (both fields would be zero values for it).

Reply:
It is just a placeholder for Go JSON parsing. I would like to start with it (as I had issues with this in the past), and once it is working, I can see if I can safely remove it.

Reply:
Ahh, yeah, I see. I looked into wasmvm, where the error case must not be empty. An empty Ok is treated as ok.

Anyways, good to have something in the field for debugging.

Reply:
> Ahh, yeah, I see. I looked into wasmvm, where the error case must not be empty. An empty Ok is treated as ok.

Ah, I guess my memory did not serve correctly here. Thanks for checking in. I will then try with `ContractResult<()>`